File: src/prepare_handler.cpp
/*
  Copyright (c) 2014-2015 DataStax

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

#include "prepare_handler.hpp"

#include "batch_request.hpp"
#include "constants.hpp"
#include "error_response.hpp"
#include "execute_request.hpp"
#include "prepare_request.hpp"
#include "request_handler.hpp"
#include "response.hpp"
#include "result_response.hpp"

namespace cass {

bool PrepareHandler::init(const std::string& prepared_id) {
  PrepareRequest* prepare =
      static_cast<PrepareRequest*>(new PrepareRequest());
  request_.reset(prepare);
  if (request_handler_->request()->opcode() == CQL_OPCODE_EXECUTE) {
    const ExecuteRequest* execute = static_cast<const ExecuteRequest*>(
        request_handler_->request());
    prepare->set_query(execute->prepared()->statement());
    return true;
  } else if (request_handler_->request()->opcode() == CQL_OPCODE_BATCH) {
    const BatchRequest* batch = static_cast<const BatchRequest*>(
        request_handler_->request());
    std::string prepared_statement;
    if (batch->prepared_statement(prepared_id, &prepared_statement)) {
      prepare->set_query(prepared_statement);
      return true;
    }
  }
  return false; // Invalid request type
}

void PrepareHandler::on_set(ResponseMessage* response) {
  switch (response->opcode()) {
    case CQL_OPCODE_RESULT: {
      ResultResponse* result =
          static_cast<ResultResponse*>(response->response_body().get());
      if (result->kind() == CASS_RESULT_KIND_PREPARED) {
        request_handler_->retry(RETRY_WITH_CURRENT_HOST);
      } else {
        request_handler_->retry(RETRY_WITH_NEXT_HOST);
      }
    } break;
    case CQL_OPCODE_ERROR:
      request_handler_->retry(RETRY_WITH_NEXT_HOST);
      break;
    default:
      break;
  }
}

void PrepareHandler::on_error(CassError code, const std::string& message) {
  request_handler_->retry(RETRY_WITH_NEXT_HOST);
}

void PrepareHandler::on_timeout() {
  request_handler_->retry(RETRY_WITH_NEXT_HOST);
}

} // namespace cass

File: gtests/src/integration/scassandra/server/server/src/main/scala/org/scassandra/server/actors/PrepareHandler.scala
/*
 * Copyright (C) 2016 Christopher Batey and Dogan Narinc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.scassandra.server.actors

import akka.actor.Actor
import com.google.common.annotations.VisibleForTesting
import org.scassandra.codec._
import org.scassandra.codec.messages.{PreparedMetadata, RowMetadata}
import org.scassandra.server.actors.PrepareHandler.{PreparedStatementQuery, PreparedStatementResponse}
import org.scassandra.server.priming._
import org.scassandra.server.priming.prepared.PreparedStoreLookup
import org.scassandra.server.priming.query.{Fatal, Reply}
import scodec.bits.ByteVector


class PrepareHandler(primePreparedStore: PreparedStoreLookup, activityLog: ActivityLog) extends ProtocolActor {

  private var nextId: Int = 1
  private var idToStatement: Map[Int, (String, Prepared)] = Map()

  @VisibleForTesting
  val generatePrepared = (p: PreparedMetadata, r: RowMetadata) => {
    val prepared = Prepared(ByteVector(nextId), preparedMetadata = p, resultMetadata = r)
    nextId += 1
    prepared
  }

  def receive: Actor.Receive = {
    case ProtocolMessage(Frame(header, p: Prepare)) =>
      activityLog.recordPreparedStatementPreparation(PreparedStatementPreparation(p.query))
      handlePrepare(header, p)
    case PreparedStatementQuery(ids) =>
      sender() ! PreparedStatementResponse(ids.flatMap(id => idToStatement.get(id) match {
        case Some(p) => Seq(id -> p)
        case None => Seq()
      }) toMap)
  }

  private def handlePrepare(header: FrameHeader, prepare: Prepare) = {
    val prime = primePreparedStore(prepare, generatePrepared)
      .getOrElse(PreparedStoreLookup.defaultPrepared(prepare, generatePrepared))

    prime match {
      case Reply(p: Prepared, _, _) =>
        idToStatement += (p.id.toInt() -> (prepare.query, p))
        log.info(s"Prepared Statement has been prepared: |$prepare.query|. Prepared result is: $p")
      case Reply(m: Message, _, _) =>
        log.info(s"Got non-prepared response for query: |$prepare.query|. Result was: $m")
      case f: Fatal =>
        log.info(s"Got non-prepared response for query: |$prepare.query|. Result was: $f")
    }

    writePrime(prepare, Some(prime), header)
  }
}

object PrepareHandler {
  case class PreparedStatementQuery(id: List[Int])
  case class PreparedStatementResponse(prepared: Map[Int, (String, Prepared)])
}